### PR TITLE
Prefs [TASK][28979] Create repository

### PR DESCRIPTION
**Pull request by viet06061999**
## Related Tickets
- [Task#28979: Create repository](https://edu-redmine.sun-asterisk.vn/issues/28979)

## WHAT
Create repository
## HOW
Create interface repository and implement
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/travelapp/data/repository/RestaurantRepository.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+        fun searchRestaurantsByProperty(
+            prameters: Map<String, String>,
+            callback: OnDataCallback<List<Restaurant>>
+        )
+        fun getDetailRestaurant(
+            prameters: Map<String, String>,
+            callback: OnDataCallback<Restaurant>
+        )
+        fun insertLocation(restaurant: Restaurant, callback: OnDataCallback<Boolean>)
+        fun getDefaultParams(): Map<String, String>
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/data/repository/RestaurantRepository.kt]: 1. sai format, thụt lề 4 spaces thôi
2. paramters kia là gì thế, getDetail Restaurant thì truyền vào cái gì thế

Comment (viet06061999) [reply app/src/main/java/com/sunasterisk/travelapp/data/repository/RestaurantRepository.kt]: cái api này nó có 1 số param để request ạ, ví dụ như language, currency, location_id của cái Restaurant đấy,

File: app/src/main/java/com/sunasterisk/travelapp/di/Injector.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        val database = DatabaseHelper.getInstance(DB_NAME, DB_VERSION, context)
+        val locationDAO = LocationFavouriteDAOImpl.getInstance(database)
+        val pref = PreferencesHelperImpl.getInstance(context)
+        val local = HotelLocalDatasource.getInstance(locationDAO, pref)
+        val remote = HotelRemoteDatasource.getInstance()
+        return HotelRepositoryImpl.getInstance(local, remote)
+    }
+
+    fun getRestaurantRepository(context: Context): RestaurantRepository {
+        val database = DatabaseHelper.getInstance(DB_NAME, DB_VERSION, context)
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/di/Injector.kt]: những cái thuộc tính DB_NAME, DB_VERSION thì chỉ cần để private trong DatabaseHelper là được 
tham số truyền vào chỉ cần Context
Thứ tự các paramters thì context luôn được ưu tiên đứng đầu 
```suggestion
        val database = DatabaseHelper.getInstance(context)
```

File: app/src/main/java/com/sunasterisk/travelapp/data/source/local/database/DatabaseHelper.kt [hunk 3]
```diff
@@ -10,7 +10,6 @@ import com.sunasterisk.travelapp.data.models.User
 
 class DatabaseHelper private constructor(dbName: String, version: Int, context: Context) :
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/data/source/local/database/DatabaseHelper.kt]:
```suggestion
class DatabaseHelper private constructor(context: Context, dbName: String, version: Int) :
```

File: app/src/main/java/com/sunasterisk/travelapp/data/repository/HotelRepository.kt [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package com.sunasterisk.travelapp.data.repository
+
+import com.sunasterisk.travelapp.data.OnDataCallback
+import com.sunasterisk.travelapp.data.models.Hotel
+
+interface HotelRepository {
+    fun searchHotelsByProperty(
+        prameters: Map<String, String>,
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/data/repository/HotelRepository.kt]: những cái param này em viết ra cho nó clear 
```suggestion
        localId: String,
        restaurantId: String,
        ....
```

Comment (huytq-2422) [reply app/src/main/java/com/sunasterisk/travelapp/data/repository/HotelRepository.kt]: Vì tầng logic bên trên không cần biết các đường path của api

File: app/src/main/java/com/sunasterisk/travelapp/data/repository/RestaurantRepositoryImpl.kt [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import com.sunasterisk.travelapp.utils.ApiEndpoint.RESTAURANT_TYPE
+import com.sunasterisk.travelapp.utils.KeyResponse.DATA
+import org.json.JSONException
+import org.json.JSONObject
+
+class RestaurantRepositoryImpl private constructor(
+    private val local: RestaurantDatasource.Local,
+    private val remote: RestaurantDatasource.Remote
+) : RestaurantRepository {
+    override fun searchRestaurantsByProperty(
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/data/repository/RestaurantRepositoryImpl.kt:18]: Để cách 1 dòng trống phân biệt header và body

File: app/src/main/java/com/sunasterisk/travelapp/data/repository/RestaurantRepositoryImpl.kt [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        latitude: String,
+        longtitude: String,
+        mealType: String?,
+        restaurantType: String?,
```
Comment (huytq-2422) [inline app/src/main/java/com/sunasterisk/travelapp/data/repository/RestaurantRepositoryImpl.kt:22]: gán giá trị mặc định 
```suggestion
        mealType: String? = null,
        restaurantType: String? = null,
```

**Pull request merged**